### PR TITLE
Better mark_dynamic assertions

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #95640
* #95639
* #95584
* __->__ #95566

This PR allows us to reuse the static per tensor decision making we make at fake tensorification time. We can use this to avoid setting up dynamic dim guards later if the tensor was never a candidate. 

cc @soumith @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

File: torch/_dynamo/utils.py [hunk 1]
```diff
@@ -92,6 +92,7 @@ def profile_wrapper(*args, **kwargs):
 
 curr_frame = 0
 
+
```
Comment (voznesenskym) [inline torch/_dynamo/utils.py:95]: I promise the linter did this lol

File: torch/_dynamo/utils.py [hunk 2]
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
...
+# the util feels worth not replicating this logic in multiple places,
+# even if the return type is really ugly.
+def tensor_shape_should_be_static(
+    tensor: Union[torch.Tensor, Any], source: Optional["Source"], is_tensor: bool
+) -> Tuple[bool, str]:
+    """
+    Given a tensor, source, and is_tensor flag, determine if a shape should be static.
+
+    Args:
+    tensor - the tensor to evaluate, parameters force a static shape.
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1340]: You should specify if this has to be real tensor, or if fake tensor is allowed

File: torch/_dynamo/utils.py [hunk 3]
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
...
+    source - an optional source, None forces a static shape
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1341]: Do you remember when `wrap_to_fake_tensor_and_record` can be given a None source?

File: test/dynamo/test_misc.py [hunk 4]
```diff
@@ -4578,6 +4578,35 @@ def my_dyn_fn(x):
...
+        e = encoder(y)
+        torch._dynamo.optimize("eager")(e)(x)
+        torch._dynamo.mark_dynamic(y, 0)
+        torch._dynamo.reset()
+        e = encoder(y)
+        with self.assertRaisesRegex(
+            AssertionError,
+            "mark_dynamic on parameter self.param, parameters are always static today",
+        ):
+            torch._dynamo.optimize("eager")(e)(x)
```
Comment (ezyang) [inline test/dynamo/test_misc.py:4608]: What exactly is this test testing? Isn't this a semantics preserving refactor?

File: torch/_dynamo/utils.py [hunk 5]
```diff
@@ -1325,3 +1325,37 @@ def get_custom_getattr(value: Any):
...
+
+    NOTE - this util ALSO returns a string, used in case of a mark_dynamic assertion going
+    against this logic.
+    """
+    if source is None:
+        return True, "mark_dynamic usage without a source is illegal."
+    if type(tensor) is torch.nn.Parameter:
+        return (
+            True,
+            f"mark_dynamic on parameter {source.name()}, parameters are always static today.",
```
Comment (ezyang) [inline torch/_dynamo/utils.py:1355]: nit: consider returning a lambda instead, otherwise you always pay for string formatting even if string message is not used

Comment (ezyang) [reply torch/_dynamo/utils.py:1355]: Alternately, a more factored representation would be to return an enum from this function saying, essentially "why is this tensor static", and then have a separately printing function for that case. That also fixes the ugly `Tuple[bool, str]` signature which I also agree is ugly.

File: test/dynamo/test_misc.py (hunk 4 above)
Comment (ezyang) [reply test/dynamo/test_misc.py:4608]: better error message

**Review by ezyang — APPROVED** [review]
Please see comments

File: torch/_dynamo/utils.py (hunk 5 above)
Comment (voznesenskym) [reply torch/_dynamo/utils.py:1355]: Enum sounds nice

File: torch/_dynamo/utils.py (hunk 3 above)
Comment (voznesenskym) [reply torch/_dynamo/utils.py:1341]: No, let me check

Comment (voznesenskym) [reply torch/_dynamo/utils.py:1357]: stacked a PR on top, I didn't see anything in tests, but there may be some part of CI that triggers it.

Comment (ezyang) [reply torch/_dynamo/utils.py:1357]: may i recommend good old fashioned code reading
